shared/trust/hmac.go

```diff
@@ -0,0 +1,265 @@
+// Example: `Authorization: <version> <format including the HMAC>`.
+type HMACVersion string
+
+// HMACConf represents the HMAC configuration.
+type HMACConf struct {
+	HashFunc   func() hash.Hash
+	Passphrase []byte
+	Version    HMACVersion
+}
+
+// HMACFormat represents arbitrary formats to diplay and parse the actual HMAC.
+// Implementations like argon extend the format with an additional salt.
+// The Header function accepts the used version and the HMAC computed over the payload.
+// It is expected to return the final Authorization header.
+// The ParseHeader function accepts a requests header and is expected to return
+// the used version and the HMAC.
+// The Passphrase function is invoked so that the format can learn about the secret
+// used for generating the HMAC and can apply modifications such as running it through a KDF.
+// Example using argon2: `Authorization: <version> <salt>:<HMAC>`.
+type HMACFormat interface {
+	Passphrase(passphrase []byte) []byte
+	Header(version HMACVersion, hmac []byte) string
+	ParseHeader(header string) (HMACVersion, []byte, error)
+}
+
+// HMAC represents the the tooling for creating and validating HMACs.
+type HMAC struct {
+	conf   HMACConf
+	format HMACFormat
+}
+
+// HMACFormatDefault represents the default format.
+type HMACFormatDefault struct{}
+
+// HMACFormatArgon represents the argon2 format.
+type HMACFormatArgon struct {
+	Salt      []byte
+	Time      uint32
+	Memory    uint32
+	Threads   uint8
+	KeyLength uint32
+}
+
+// NewDefaultHMACConf returns the default configuration for HMAC.
+func NewDefaultHMACConf(passphrase string, version HMACVersion) HMACConf {
+	return HMACConf{
+		Passphrase: []byte(passphrase),
+		HashFunc:   sha256.New,
+		Version:    version,
+	}
+}
+
+// Passphrase accepts the outer HMAC passphrase and returns the same without modification.
+func (h *HMACFormatDefault) Passphrase(passphrase []byte) []byte {
+	return passphrase
+}
+
+// Header returns the actual HMAC alongside together with the used version.
+func (h *HMACFormatDefault) Header(version HMACVersion, hmac []byte) string {
+	return fmt.Sprintf("%s %s", version, hex.EncodeToString(hmac))
+}
+
+// ParseHeader extracts the actual version and HMAC from the header.
+func (h *HMACFormatDefault) ParseHeader(header string) (HMACVersion, []byte, error) {
+	authHeaderSplit := strings.Split(header, " ")
+	if len(authHeaderSplit) != 2 {
+		return "", nil, errors.New("Version or HMAC is missing")
+	}
+
+	hmac, err := hex.DecodeString(authHeaderSplit[1])
+	if err != nil {
+		return "", nil, fmt.Errorf("Failed to decode the HMAC: %w", err)
+	}
+
+	return HMACVersion(authHeaderSplit[0]), hmac, nil
+}
+
+// NewDefaultHMACFormatArgon returns the default format for argon2.
+// Recommended defaults according to https://www.rfc-editor.org/rfc/rfc9106#section-4-6.2.
+// We use the second recommended option to not require a system having 2 GiB of memory.
+func NewDefaultHMACFormatArgon() (*HMACFormatArgon, error) {
```

Comment:
Can this take a salt argument which can avoid generating a random one if provided?